sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,340 @@
+        The categories of each feature determined during fitting
+        (in order of the features in `X` and corresponding with the output
+        of :meth:`transform`).
+
+    target_type_ : str
+        Type of target.
+
+    target_mean_ : float
+        The overall mean of the target. This value is only used in :meth:`transform`
+        to encode categories.
+
+    n_features_in_ : int
+        Number of features seen during :term:`fit`.
+
+    feature_names_in_ : ndarray of shape (`n_features_in_`,)
+        Names of features seen during :term:`fit`. Defined only when `X`
+        has feature names that are all strings.
+
+    See Also
+    --------
+    OrdinalEncoder : Performs an ordinal (integer) encoding of the categorical features.
+        Contrary to TargetEncoder, this encoding is not supervised. Treating the
+        resulting encoding as a numerical features therefore lead arbitrarily
+        ordered values and therefore typically lead to lower predictive performance
+        when used as preprocessing for a classifier or regressor.
+    OneHotEncoder : Performs a one-hot encoding of categorical features. This
+        unsupervised encoding is better suited for low cardinality categorical
+        variables as it generate one new feature per unique category.
+
+    References
+    ----------
+    .. [MIC] :doi:`Micci-Barreca, Daniele. "A preprocessing scheme for high-cardinality
+       categorical attributes in classification and prediction problems"
+       SIGKDD Explor. Newsl. 3, 1 (July 2001), 27–32. <10.1145/507533.507538>`
+
+    Examples
+    --------
+    With `smooth="auto"`, the smoothing parameter is set to an empirical Bayes estimate:
+    >>> import numpy as np
+    >>> from sklearn.preprocessing import TargetEncoder
+    >>> X = np.array([["dog"] * 20 + ["cat"] * 30 + ["snake"] * 38], dtype=object).T
+    >>> y = [90.3] * 5 + [80.1] * 15 + [20.4] * 5 + [20.1] * 25 + [21.2] * 8 + [49] * 30
+    >>> enc_auto = TargetEncoder(smooth="auto")
+    >>> X_trans = enc_auto.fit_transform(X, y)
+
+    >>> # A high `smooth` parameter puts more weight on global mean on the categorical
+    >>> # encodings:
+    >>> enc_high_smooth = TargetEncoder(smooth=5000.0).fit(X, y)
+    >>> enc_high_smooth.target_mean_
+    44...
+    >>> enc_high_smooth.encodings_
+    [array([44..., 44..., 44...])]
+
+    >>> # On the other hand, a low `smooth` parameter puts more weight on target
+    >>> # conditioned on the value of the categorical:
+    >>> enc_no_smooth = TargetEncoder(smooth=1.0).fit(X, y)
+    >>> enc_no_smooth.encodings_
+    [array([20..., 80..., 43...])]
+    """
+
+    _parameter_constraints: dict = {
+        "categories": [StrOptions({"auto"}), list],
+        "target_type": [StrOptions({"auto", "continuous", "binary"})],
+        "smooth": [StrOptions({"auto"}), Interval(Real, 0, None, closed="left")],
+        "cv": [Interval(Integral, 2, None, closed="left")],
```

Comment:
Shall we ally to pass `cv=None` to disable internal CV? This could be useful for educational purposes and statistically study the impact of the nested CV.

Reply:
As discussed at the triage meeting, we can actually manually investigate the impact of disabling the cross-validation by manually calling `.fit(X_train).transform(X_train)` in our own code snippets. But this prevents using a pipeline to study the impact of the internal cross-validation. But maybe this is enough.

Reply:
Looking at this again, I am +0.3 with adding `cv=None` for educational purposes. To include `cv=None`, I would want more docs explaining why `cv=None` is normally bad to do. Given the current scope of this PR, can we push `cv=None` for a follow up PR?

I recall you wanted to add a test with `cv=None`. In that case, we can have a "hidden" `None` option that disables the CV, used only for testing purposes.

Reply:
Yes. Let‘s keep this PR scoped and postpone `cv=None` to a possible future PR.

Reply:
I am also +0.42 for adding `cv=None` and +1 for doing this in a follow-up PR along with an (updated or dedicated) example that document the pitfall of using `cv=None`.

Reply:
If we implement `cv=None` we can make the test in 26d2429 more natural.